src/dparse/parser.d

```diff
@@ -2864,29 +2864,32 @@ class Parser
      * Parses a ForeachType
      *
      * $(GRAMMAR $(RULEDEF foreachType):
-     *       $(LITERAL 'ref')? $(RULE typeConstructors)? $(RULE type)? $(LITERAL Identifier)
-     *     | $(RULE typeConstructors)? $(LITERAL 'ref')? $(RULE type)? $(LITERAL Identifier)
+     *       ($(LITERAL 'ref') | $(LITERAL 'alias') | $(LITERAL 'enum') | $(RULE typeConstructor))* $(RULE type)? $(LITERAL Identifier)
      *     ;)
      */
     ForeachType parseForeachType()
     {
         mixin(traceEnterAndExit!(__FUNCTION__));
         auto node = allocator.make!ForeachType;
-        if (currentIs(tok!"ref"))
-        {
-            node.isRef = true;
-            advance();
-        }
-        if (currentIsOneOf(tok!"const", tok!"immutable",
-            tok!"inout", tok!"shared") && !peekIs(tok!"("))
-        {
-            trace("\033[01;36mType constructor");
-            mixin(parseNodeQ!(`node.typeConstructors`, `TypeConstructors`));
-        }
-        if (currentIs(tok!"ref"))
+        while (moreTokens())
         {
-            node.isRef = true;
-            advance();
+            IdType typeConstructor;
+            if (currentIs(tok!"ref"))
+            {
+                node.isRef = true;
+                advance();
+            }
+            else if (currentIs(tok!"alias"))
+                advance(); // TODO: where should I put this
```

Comment:
I'm not sure where the new attributes should be added. Should I make a new field `IdType[] attributes`? if so should it include ref? should it just be an alias to `typeConstructors` field? Should I add two more flags `bool isEnum` and `bool isAlias`? Or should it be a `IdType enumOrAlias` field?

Reply:
That's a good question. Adding the extra isEnum and isAlias fields would mean that there aren't any backwards-incompatible changes to the AST classes, so I think that's the better thing to do even thought it might not result in the cleanest design.